cf-agent/verify_packages.c

```diff
@@ -131,8 +144,91 @@ PackageManager *INSTALLED_PACKAGE_LISTS = NULL; /* GLOBAL_X */
 
 #define PACKAGE_IGNORED_CFE_INTERNAL "cfe_internal_non_existing_package"
 
+/* Returns the old or new package promise type depending on promise 
+   constraints. */
+static PackagePromiseType GetPackagePromiseVersion(const Packages *packages,
+        const NewPackages *new_packages)
+{
+    /* We have mixed packages promise constraints. */
+    if (!packages->is_empty && !new_packages->is_empty)
+    {
+        return PACKAGE_PROMISE_TYPE_MIXED;
+    }
+    else if (!new_packages->is_empty) /* new packages promise */
+    {
+        if (new_packages->package_policy == NEW_PACKAGE_ACTION_NONE)
+        {
+            return PACKAGE_PROMISE_TYPE_NEW_ERROR;
+        }
+        return PACKAGE_PROMISE_TYPE_NEW;
+    }
+    else /* old packages promise */
+    {
+        //TODO:
+        if (!packages->has_package_method)
+        {
+            return PACKAGE_PROMISE_TYPE_OLD_ERROR;
+        }
+        return PACKAGE_PROMISE_TYPE_OLD;
+    }
+}
+
+PromiseResult VerifyPackagesPromise(EvalContext *ctx, const Promise *pp)
+{
+    PromiseResult result = PROMISE_RESULT_FAIL;
+    char *promise_log_message;
+    LogLevel level;
+    
+    Attributes a = GetPackageAttributes(ctx, pp);
+    PackagePromiseType package_promise_type =
+            GetPackagePromiseVersion(&a.packages, &a.new_packages);
+    
+    switch (package_promise_type)
+    {
+        case PACKAGE_PROMISE_TYPE_NEW:
+            result = HandleNewPackagePromiseType(ctx, pp, a, &promise_log_message,
+                    &level);
+            
```

Comment:
I would initialise `promise_log_message = NULL` before this call, and here I would `assert(promise_log_message != NULL)`. It was hard to follow the logic of the function call, and I bet it will be easy to regress after some time when someone forgets to set the log message.